server/buffer.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package server
+
+import (
+	"time"
+
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/topdown"
+)
+
+// Buffer defines the interface for types that can be used as the diagnostic buffer
+// within the OPA server.
+type Buffer interface {
+	Push(*Info)
+	Pop() (*Info, bool)
+	Iter(fn func(*Info))
+}
+
+// buffer stores diagnostic information. They are not safe for concurrent access.
+type buffer struct {
+	ring []*Info
+
+	size int
+
+	start int // The index of the next item to pop.
+	end   int // The index where the next item will be inserted.
+}
+
+// NewBoundedBuffer creates a new Buffer with maximum size n. NewBoundedBuffer will panic if n is not
+// positive.
+func NewBoundedBuffer(n int) Buffer {
+	if n <= 0 {
+		panic("size must be greater than 0")
+	}
+
+	return &buffer{ring: make([]*Info, n, n)}
+}
+
+// Push inserts the provided Info into the buffer. If the buffer is full, the oldest item
+// is deleted to make room.
+func (b *buffer) Push(i *Info) {
+	b.ring[b.end] = i
+
+	b.end = (b.end + 1) % len(b.ring)
+
+	switch b.size {
+	case len(b.ring): // We erased something, move the start up.
+		b.start = (b.start + 1) % len(b.ring)
+	default:
+		b.size++
+	}
+}
+
+// Pop removes the oldest item from the buffer and returns it. If the buffer is empty, ok
+// will be false.
+func (b *buffer) Pop() (i *Info, ok bool) {
+	if b.size == 0 {
+		return nil, false
+	}
+
+	i = b.ring[b.start]
+	b.ring[b.start] = nil // Free it up for garbage collection.
+
+	b.start = (b.start + 1) % len(b.ring)
+	b.size--
+
+	return i, true
+}
+
+// Iter iterates over the buffer, calling fn for every element in it.
+func (b *buffer) Iter(fn func(*Info)) {
+	i := b.start
+	for j := 0; j < b.size; j++ {
+		fn(b.ring[i])
+		i = (i + 1) % len(b.ring)
+	}
+}
+
+// Info stores diagnostic information about the evaluation of a query.
+type Info struct {
+	query     string
+	timestamp time.Time
+
+	input   interface{}
+	results interface{}
+
+	metrics metrics.Metrics
+	trace   *topdown.BufferTracer
```

Comment:
My last comment on this got lost in the move, so restating here...

The diagnostics should store a trace -- not the tracer itself.